### PR TITLE
NAS-114146 / 22.02 / Add support for "duplicate_serial" field

### PR DESCRIPTION
**Pull request by denysbutenko**
Display warning if available disks have items in `duplicate_serial`

Query to mock `disk.get_unused` response in Shell:
```
midclt call test.set_mock disk.get_unused '{"return_value": [{"identifier": "{uuid}07c19ad9-5b8d-4ad5-b7bc-9a3a8ce50c9e", "name": "sda", "subsystem": "scsi", "number": 1, "serial": "", "lunid": null, "size": 21474836480, "description": "", "transfermode": "Auto", "hddstandby": "ALWAYS ON", "advpowermgmt": "DISABLED", "togglesmart": true, "smartoptions": "", "expiretime": null, "critical": null, "difference": null, "informational": null, "model": "VMware_Virtual_S", "rotationrate": null, "type": "HDD", "zfs_guid": null, "bus": "SPI", "devname": "sda", "enclosure": null, "pool": null, "duplicate_serial": ["sdc", "sdb", "sdd", "sde", "sdf"]}, {"identifier": "{devicename}sdb", "name": "sdb", "subsystem": "scsi", "number": 2, "serial": "", "lunid": null, "size": 21474836480, "description": "", "transfermode": "Auto", "hddstandby": "ALWAYS ON", "advpowermgmt": "DISABLED", "togglesmart": true, "smartoptions": "", "expiretime": null, "critical": null, "difference": null, "informational": null, "model": "VMware_Virtual_S", "rotationrate": null, "type": "HDD", "zfs_guid": "5383752207637265814", "bus": "SPI", "devname": "sdb", "enclosure": null, "pool": null, "duplicate_serial": ["sdc", "sda", "sdd", "sde", "sdf"]}, {"identifier": "{devicename}sdd", "name": "sdd", "subsystem": "scsi", "number": 4, "serial": "", "lunid": null, "size": 21474836480, "description": "", "transfermode": "Auto", "hddstandby": "ALWAYS ON", "advpowermgmt": "DISABLED", "togglesmart": true, "smartoptions": "", "expiretime": null, "critical": null, "difference": null, "informational": null, "model": "VMware_Virtual_S", "rotationrate": null, "type": "HDD", "zfs_guid": "5995531091461019479", "bus": "SPI", "devname": "sdd", "enclosure": null, "pool": null, "duplicate_serial": ["sdc", "sda", "sdb", "sde", "sdf"]}, {"identifier": "{devicename}sde", "name": "sde", "subsystem": "scsi", "number": 5, "serial": "", "lunid": null, "size": 21474836480, "description": "", "transfermode": "Auto", "hddstandby": "ALWAYS ON", "advpowermgmt": "DISABLED", "togglesmart": true, "smartoptions": "", "expiretime": null, "critical": null, "difference": null, "informational": null, "model": "VMware_Virtual_S", "rotationrate": null, "type": "HDD", "zfs_guid": null, "bus": "SPI", "devname": "sde", "enclosure": null, "pool": null, "duplicate_serial": ["sdc", "sda", "sdb", "sdd", "sdf"]}, {"identifier": "{devicename}sdf", "name": "sdf", "subsystem": "scsi", "number": 6, "serial": "", "lunid": null, "size": 21474836480, "description": "", "transfermode": "Auto", "hddstandby": "ALWAYS ON", "advpowermgmt": "DISABLED", "togglesmart": true, "smartoptions": "", "expiretime": null, "critical": null, "difference": null, "informational": null, "model": "VMware_Virtual_S", "rotationrate": null, "type": "HDD", "zfs_guid": null, "bus": "SPI", "devname": "sdf", "enclosure": null, "pool": null, "duplicate_serial": ["sdc", "sda", "sdb", "sdd", "sde"]}] }'
```

Query for testing USB message `bus: 'USB'` :
```
midclt call test.set_mock disk.get_unused '{"return_value": [{"identifier": "{uuid}07c19ad9-5b8d-4ad5-b7bc-9a3a8ce50c9e", "name": "sda", "subsystem": "scsi", "number": 1, "serial": "", "lunid": null, "size": 21474836480, "description": "", "transfermode": "Auto", "hddstandby": "ALWAYS ON", "advpowermgmt": "DISABLED", "togglesmart": true, "smartoptions": "", "expiretime": null, "critical": null, "difference": null, "informational": null, "model": "VMware_Virtual_S", "rotationrate": null, "type": "HDD", "zfs_guid": null, "bus": "USB", "devname": "sda", "enclosure": null, "pool": null, "duplicate_serial": ["sdc", "sdb", "sdd", "sde", "sdf"]}, {"identifier": "{uuid}316eaa17-1722-40a3-833f-066f76ee2948", "name": "sdb", "subsystem": "scsi", "number": 2, "serial": "", "lunid": null, "size": 21474836480, "description": "", "transfermode": "Auto", "hddstandby": "ALWAYS ON", "advpowermgmt": "DISABLED", "togglesmart": true, "smartoptions": "", "expiretime": null, "critical": null, "difference": null, "informational": null, "model": "VMware_Virtual_S", "rotationrate": null, "type": "HDD", "zfs_guid": null, "bus": "USB", "devname": "sdb", "enclosure": null, "pool": null, "duplicate_serial": ["sdc", "sda", "sdd", "sde", "sdf"]}, {"identifier": "{uuid}d64d1d72-05e8-46ab-a478-bc9b7ff73067", "name": "sdd", "subsystem": "scsi", "number": 4, "serial": "", "lunid": null, "size": 21474836480, "description": "", "transfermode": "Auto", "hddstandby": "ALWAYS ON", "advpowermgmt": "DISABLED", "togglesmart": true, "smartoptions": "", "expiretime": null, "critical": null, "difference": null, "informational": null, "model": "VMware_Virtual_S", "rotationrate": null, "type": "HDD", "zfs_guid": null, "bus": "USB", "devname": "sdd", "enclosure": null, "pool": null, "duplicate_serial": ["sdc", "sda", "sdb", "sde", "sdf"]}, {"identifier": "{uuid}e44343aa-1235-43bf-9371-ece8180ca2c9", "name": "sde", "subsystem": "scsi", "number": 5, "serial": "", "lunid": null, "size": 21474836480, "description": "", "transfermode": "Auto", "hddstandby": "ALWAYS ON", "advpowermgmt": "DISABLED", "togglesmart": true, "smartoptions": "", "expiretime": null, "critical": null, "difference": null, "informational": null, "model": "VMware_Virtual_S", "rotationrate": null, "type": "HDD", "zfs_guid": null, "bus": "USB", "devname": "sde", "enclosure": null, "pool": null, "duplicate_serial": ["sdc", "sda", "sdb", "sdd", "sdf"]}] }'
```

Testing:
- Use the latest nightly on a VM
- Mock response or create new disks
- Go to `Storage - Create Pool`
- Check the warning message
- Toggle Show/Hide all disks
- Add disk with non-unique serial to vdev
- Try to break form submission

Also check ticket details for additional info

### CODE REVIEW

**Review by RehanY147 — CHANGES_REQUESTED** [review]
The text should be changed to
`Warning: There are {n} disks available that have non-unique serial numbers. Non-unique serial numbers can be caused by a cabling issue and adding such disks to a pool can result in lost data.`

`Warning: There are {n} USB disks available that have non-unique serial numbers. USB controllers may report disk serial incorrectly, making such disks indistinguishable from each other.`

Also, instead of a link, we should have a checkbox there with a message like
`Show disks with non-unique serial numbers` and `Hide disks with non-unique serial numbers`

The link is not really distinguishable from the text and makes it difficult to understand an action can be performed there.

File: src/app/pages/storage/volumes/manager/manager.component.ts [hunk 1]
```diff
@@ -126,6 +128,26 @@ export class ManagerComponent implements OnInit, AfterViewInit {
...
+    }
+    return this.temp.filter((disk) => this.nonUniqueSerialDisks.includes(disk)).length;
+  }
+  get nonUniqueSerialDisksWarning(): string {
+    if (!this.nonUniqueSerialDisks.length) {
+      return null;
+    }
+
+    if (this.nonUniqueSerialDisks.every((disk) => disk.bus === DiskBus.Usb)) {
+      return this.translate.instant('Warning: There are {n} USB disks available that have non-unique serial numbers. USB controllers may report disk serial incorrectly, making such disks indistinguishable from each other.', { n: this.availableNonUniqueSerialDisksCount });
```
Comment (RehanY147) [inline src/app/pages/storage/volumes/manager/manager.component.ts:145]: If data loss is possible in this case too, we should mention that here as well. `Adding such disks to a pool can result in lost data.` or something like that.

**Review by RehanY147 — CHANGES_REQUESTED** [review]

File: src/app/pages/storage/volumes/manager/manager.component.ts [hunk 2]
```diff
@@ -126,6 +128,26 @@ export class ManagerComponent implements OnInit, AfterViewInit {
     stripe: 1, mirror: 2, raidz: 3, raidz2: 4, raidz3: 5,
   };
 
+  includeNonUniqueSerialDisks = false;
+  nonUniqueSerialDisks: ManagerDisk[] = [];
+  get availableNonUniqueSerialDisksCount(): number {
+    if (this.originalDisks.length === this.temp.length) {
+      return this.nonUniqueSerialDisks.length;
+    }
+    return this.temp.filter((disk) => this.nonUniqueSerialDisks.includes(disk)).length;
```
Comment (RehanY147) [inline src/app/pages/storage/volumes/manager/manager.component.ts:137]: We should rename `temp` to something more readable

**Review by RehanY147 — APPROVED** [review]
The code looks good and works as intended

File: src/app/pages/storage/volumes/manager/manager.component.ts (hunk 2 above)
Comment (denysbutenko) [reply src/app/pages/storage/volumes/manager/manager.component.ts:137]: Your suggestions?

**Pull request merged**